ouroboros/chainsync/v5/types.go

```diff
@@ -0,0 +1,656 @@
+		var ps PointStructV5
+		if err := json.Unmarshal(data, &ps); err != nil {
+			return fmt.Errorf("failed to unmarshal Point, %v: %w", string(data), err)
+		}
+
+		*p = PointV5{
+			pointType:   chainsync.PointTypeStruct,
+			pointStruct: &ps,
+		}
+	}
+
+	return nil
+}
+
+type ResultV5 struct {
+	IntersectionFound    *IntersectionFound    `json:",omitempty" dynamodbav:",omitempty"`
+	IntersectionNotFound *IntersectionNotFound `json:",omitempty" dynamodbav:",omitempty"`
+	RollForward          *RollForwardV5        `json:",omitempty" dynamodbav:",omitempty"`
+	RollBackward         *RollBackwardV5       `json:",omitempty" dynamodbav:",omitempty"`
+}
+
+type ResultFindIntersectionV5 struct {
+	IntersectionFound    *IntersectionFoundV5    `json:",omitempty" dynamodbav:",omitempty"`
+	IntersectionNotFound *IntersectionNotFoundV5 `json:",omitempty" dynamodbav:",omitempty"`
+}
+
+type RollBackwardV5 struct {
+	Point PointV5 `json:"point,omitempty" dynamodbav:"point,omitempty"`
+	Tip   TipV5   `json:"tip,omitempty"   dynamodbav:"tip,omitempty"`
+}
+
+type RollForwardV5 struct {
+	Block BlockV5 `json:"block,omitempty" dynamodbav:"block,omitempty"`
+	Tip   TipV5   `json:"tip,omitempty"   dynamodbav:"tip,omitempty"`
+}
+
+type ResultNextBlockV5 struct {
+	RollForward  *RollForwardV5  `json:",omitempty" dynamodbav:",omitempty"`
+	RollBackward *RollBackwardV5 `json:",omitempty" dynamodbav:",omitempty"`
+}
+
+type IntersectionFoundV5 struct {
+	Point *chainsync.Point
+	Tip   *TipV5
+}
+
+type IntersectionNotFoundV5 struct {
+	Tip *TipV5
+}
+
+// Support findIntersect (v6) / FindIntersection (v5) universally.
+type CompatibleResultFindIntersection chainsync.ResultFindIntersectionPraos
+
+func (c *CompatibleResultFindIntersection) UnmarshalJSON(data []byte) error {
+	// Assume v6 responses first, then fall back to manual v5 processing.
+	var r chainsync.ResultFindIntersectionPraos
+	err := json.Unmarshal(data, &r)
+	if err == nil && r.Tip != nil {
+		*c = CompatibleResultFindIntersection(r)
+		return nil
+	}
+
+	var r5 ResultFindIntersectionV5
+	err = json.Unmarshal(data, &r5)
+	if err != nil {
+		return err
+	} else if r5.IntersectionFound != nil {
+		tip := r5.IntersectionFound.Tip.ConvertToV6()
+		c.Intersection = r5.IntersectionFound.Point
+		c.Tip = &tip
+		c.Error = nil
+		c.ID = nil
+
+		return nil
+	} else if r5.IntersectionNotFound != nil {
+		// Emulate the v6 IntersectionNotFound error as best as possible.
+		tip := r5.IntersectionNotFound.Tip.ConvertToV6()
+		err := chainsync.ResultError{Code: 1000, Message: "Intersection not found", Data: &tip}
+		c.Error = &err
+
+		return nil
+	}
+
+	// TODO: Further error handling here.
+	return nil
+}
+
+func (c CompatibleResultFindIntersection) String() string {
+	return fmt.Sprintf("intersection=[%v] tip=[%v] error=[%v] id=[%v]", c.Intersection, c.Tip, c.Error, c.ID)
+}
+
+// Support findIntersect (v6) / FindIntersection (v5) universally.
+type CompatibleResultNextBlock chainsync.ResultNextBlockPraos
```

Comment:
I would leave the CompatibleX types in the top level; `v5` should be entirely an implementation detail hidden to the external consumer (unless they absolutely need it).

Plus, doesn't this create an awkward circular dependency among the types? the parent module needs to depend on v5, but v5 now needs to depend on the parent module for `chainsync.ResultNextBlockPraos`